### DISCUSSION TITLE
可否加入iCloud同步功能

### DISCUSSION THREAD

**Discussion by Ricadre** [imfuxiao/Hamster#44] [Ideas]
可否加入iCloud同步功能，实现全平台词库同步呢？感谢开发者的付出！

**Comment by jokerxsun** [comment]
iCloud 很有可能会导致卡顿。因为iCloud 奇葩的同步逻辑。

**Comment by g6ai** [comment]
请问v1.6.8/1.6.9的“拷贝键盘词库文件至应用”和“Rime同步”这两个功能，对于个人词库文件（比如`*.userdb.txt`）的处理的区别是什么呢？

**Reply by imfuxiao** [reply]
您好，解释这个功能先需要说明下另一个功能。

app输入方案现在可以显示在文档中了，即可以在Files内置应用中查看，也可使用外部文本编辑器编辑查看。

由于iOS系统限制，键盘不能直接使用上面文件路径。所以每次重新部署需要将应用下的方案复制到键盘可以访问的路径下。

所以键盘产生的自造词文件是保存在别的路径下的。而非app中的文档路径。

“拷贝键盘词库文件至应用”就是为了解决键盘与app的输入文档是两个路径的问题。

“RIME同步”就是rime原生的同步功能。

这里说明下，如果不先通过“拷贝键盘词库文件至应用“功能将键盘生成的词库文件拷贝出来，直接通过“RIME同步”是无法同步词库的。

词库同步推荐方案：

1. 通过iCloud：开启iCloud，点击 复制键盘词库文件到应用， 在点击 iCloud页面下的 拷贝应用文件至iCloud
2. 通过RIME的同步功能：首先设置RIME的同步路径为ICloud地址，然后点击复制键盘词库文件至应用，在点击RIME页面下的同步按钮

两个的区别在于：第一个方案是全量复制文件到iCloud中（包含输入方案），第二种是依赖RIME的同步功能，只复制了词库文件

**Reply by cactusnix** [reply]
你好, 可否将这两个操作合并成一个呢, 减少用户的心智负担.

**Reply by imfuxiao** [reply]
刚优化完这个功能，TF版本正在发版。

优化后的RIME页面有个开关，关闭后会保留键盘自造词；同时Rime同步功能也不在需要先点击拷贝词库文件至应用了，此功能内置到同步功能里了。原拷贝键盘词库至应用功能移至文件管理下面，基本上不需要在点击了。

**Reply by g6ai** [reply]
谢谢更新！请问“关闭后会保留键盘自造词”这一开关里的“键盘自造词”是指本地用户词库吗？即关闭开关后，保留本地键盘中产生的本地词库，不参与同步过程中使用sync文件夹的词库更新？

**Reply by imfuxiao** [reply]
是的，自造词就是键盘生成的词库。

关闭开关后，每次点击重新部署，键盘词库文件覆盖应用内的词库文件，所以键盘自造词不会丢失。默认是开启状态，重新部署会用应用下词库文件覆盖键盘的自造词文件，所以会丢失。

关于RIME的同步功能：

首先，RIME同步功能只针对userdb格式的自造词。即二进制格式。如果您的自造词是tabledb格式，即文本文件格式。那么RIME的同步功能对您来说无用。

其次，仓在调用RIME的同步功能前，会先复制键盘自造词文件到应用路径下。然后执行RIME同步任务。也就减免了之前的“拷贝键盘词库文件至应用”的步骤。

最后，您问题中的“sync文件夹的词库更新”，如果是指多个终端的词库文件，那么点击仓的同步，肯定会把sync下其他终端的词库合并到仓的词库下。这是RIME的机制，与仓无关。

**Reply by g6ai** [reply]
谢谢您的讲解！看来这个开关是处理重新部署的词库文件相关行为的。

**Comment by cactusnix** [comment]
关于 iCloud 还有一些疑问

1、File 应用本身的目录是什么作用，就是你前面提到的键盘本身能访问的地方是吗?
2、RIME 同步的目录和这个有什么区别呢?

以下是一些附件:
![image](https://github.com/imfuxiao/Hamster/assets/16420652/c949d370-1784-4cda-b466-828cc6b5261c)
![image](https://github.com/imfuxiao/Hamster/assets/16420652/d60baf15-b3a3-474e-a028-957aa6662fb7)

**Reply by imfuxiao** [reply]
> 关于 iCloud 还有一些疑问
> 
> 1、File 应用本身的目录是什么作用，就是你前面提到的键盘本身能访问的地方是吗? 2、RIME 同步的目录和这个有什么区别呢?
> 
> 以下是一些附件: ![image](https://user-images.githubusercontent.com/16420652/244259299-c949d370-1784-4cda-b466-828cc6b5261c.png) ![image](https://user-images.githubusercontent.com/16420652/244259424-d60baf15-b3a3-474e-a028-957aa6662fb7.png)

1. File应用本身的目录是什么作用？ 可以开放给第三方文件编辑器编辑文件。键盘本身的文件位置在开启完全访问权限后，是在AppGroup路径下，此路径只能在应用内查看，Files查看不了。
2. RIME的同步目录：默认（如果不在配置文件中设置同步目录）是在应用的Rime/sync下，Files可以查看。如果设置了则在你设置的路径下。注意:iOS有安全限制，和macos不一样，不是想设置什么位置都行。一般iOS同步目录设置都是在仓的iCloud同步目录的位置（应用内RIME页面中用，点击可复制）。
